README
django-sizefield
----------------
A file size field, stored as BigInteger and rendered with units in Bytes (KB, MB, ...)


Examples
--------
With a model like ::


    class Data(models.Model):
        path = models.FilePathField()
        size = FileSizeField()

In templates ::

    {% load sizefieldtags %}
    
    {{ data.size|filesize }}
    
*will render 12.3KB (for example)*

The model form will have a TextInput, which renders the 
value with units, and accepts values with or without units.




CHANGES
=========
CHANGELOG
=========

2.2.0 (unreleased)
==================

- Nothing changed yet.


2.1.0 (2023-06-29)
==================

- Remove usage of removed functions in Django 4 (thanks @hagbarddenstore)

2.0.0 (2020-08-13)
==================

- Dropped support for python 2
- Dropped support for Django <2
- Moved testing to GitHub Actions and tox

1.0.1 (unreleased)
==================

- Nothing changed yet.


1.0.0 (2019-06-21)
==================

- Fix support of Django 1.11/2


0.9.1 (2016-04-13)
==================

- Fix lazy loading of translation infrastructure (#15, thanks @cemiarni).


0.9 (2015-11-12)
================

- Add support for unitless values (#13, thanks @megapctr)


0.8 (2015-10-04)
================

- Add support for Python3 (thanks @naxube)


0.7 (2015-03-13)
================

- Add Zetta and Yotta support (thanks @meteozond)
- Fix bug with big values (fixes #8, thanks @meteozond)


0.6 (2014-09-25)
================

- Add Exabytes support (thanks Matt Pritchard)

0.5 (2014-06-18)
================

- Fix template tag not support None values (fixes #4)
- Support comma notation for float values (fixes #5)

0.4 (2013-08-25)
================

- Complete refactor by @perez

0.3 (2013-06-08)
================

- South support by @meteozond

0.2 (2013-05-31)
================

- Petabytes support by @meteozond

0.1 (2011-01-04)
================

- Initial working version.


sizefield/tests.py
from django.test import TestCase

from .utils import filesizeformat, parse_size
from .templatetags.sizefieldtags import filesize


class ParseRenderTest(TestCase):

    def test_render(self):

        from sizefield import utils
        utils.SIZEFIELD_FORMAT = '{value} {unit}'

        # Usual case
        self.assertEqual('123.0 B', filesizeformat(123))
        self.assertEqual('123.0 B', filesizeformat('123'))
        # Incorrect input
        self.assertRaises(ValueError, filesizeformat, (''))
        self.assertRaises(ValueError, filesizeformat, ('abc'))
        self.assertRaises(ValueError, filesizeformat, ('12 HB'))
        # Units
        self.assertEqual('1000.0 B', filesizeformat(1000))
        self.assertEqual('1023.0 B', filesizeformat(1023))
        self.assertEqual('1.0 KB', filesizeformat(1024))
        self.assertEqual('1.0 MB', filesizeformat(1024 * 1024))
        self.assertEqual('1.0 GB', filesizeformat(1024 * 1024 * 1024))
        self.assertEqual('512.0 KB', filesizeformat(1024 * 1024 * 0.5))
        self.assertEqual('307.2 KB', filesizeformat(1024 * 1024 * 0.3))
        # Decimals
        self.assertEqual('1 KB', filesizeformat(1024, decimals=0))
        self.assertEqual('1.0 KB', filesizeformat(1024, decimals=1))
        self.assertEqual('1.00 KB', filesizeformat(1024, decimals=2))
        self.assertEqual('1.000 KB', filesizeformat(1024, decimals=3))
        self.assertEqual('1.0000 KB', filesizeformat(1024, decimals=4))

        # test SIZEFIELD_FORMAT
        utils.SIZEFIELD_FORMAT = '{value}xxx {unit}'
        self.assertEqual('1xxx KB', filesizeformat(1024, decimals=0))
        self.assertEqual('1.0xxx KB', filesizeformat(1024, decimals=1))
        self.assertEqual('1.00xxx KB', filesizeformat(1024, decimals=2))
        self.assertEqual('1.000xxx KB', filesizeformat(1024, decimals=3))
        self.assertEqual('1.0000xxx KB', filesizeformat(1024, decimals=4))

        # Unambiguous unit format
        utils.SIZEFIELD_FORMAT = '{value} {unit}'
        self.assertEqual('1000.0 B', filesizeformat(1000, ambiguous_suffix=False))
        self.assertEqual('1023.0 B', filesizeformat(1023, ambiguous_suffix=False))
        self.assertEqual('1.0 KiB', filesizeformat(1024, ambiguous_suffix=False))
        self.assertEqual('1.0 MiB', filesizeformat(1024 * 1024, ambiguous_suffix=False))
        self.assertEqual('1.0 GiB', filesizeformat(1024 * 1024 * 1024, ambiguous_suffix=False))
        self.assertEqual('512.0 KiB', filesizeformat(1024 * 1024 * 0.5, ambiguous_suffix=False))
        self.assertEqual('307.2 KiB', filesizeformat(1024 * 1024 * 0.3, ambiguous_suffix=False))

        # Non binary
        utils.SIZEFIELD_FORMAT = '{value} {unit}'
        # Units
        self.assertEqual('999.0 B', filesizeformat(999, is_binary=False))
        self.assertEqual('1.0 KB', filesizeformat(1000, is_binary=False))
        self.assertEqual('1.0 MB', filesizeformat(1000 * 1000, is_binary=False))
        self.assertEqual('1.0 GB', filesizeformat(1000 * 1000 * 1000, is_binary=False))
        self.assertEqual('500.0 KB', filesizeformat(1000 * 1000 * 0.5, is_binary=False))
        self.assertEqual('305.5 KB', filesizeformat(1000 * 1000 * 0.3055, is_binary=False))
        self.assertEqual('100.0 B', filesizeformat(100, is_binary=False, ambiguous_suffix=False))
        # Decimals
        self.assertEqual('1 KB', filesizeformat(1000, decimals=0, is_binary=False))
        self.assertEqual('1.0 KB', filesizeformat(1000, decimals=1, is_binary=False))
        self.assertEqual('1.00 KB', filesizeformat(1000, decimals=2, is_binary=False))
        self.assertEqual('1.000 KB', filesizeformat(1000, decimals=3, is_binary=False))
        self.assertEqual('1.0000 KB', filesizeformat(1000, decimals=4, is_binary=False))

    def test_parse(self):
        # Usual case
        self.assertEqual(123, parse_size('123'))
        self.assertEqual(123, parse_size('123B'))
        self.assertEqual(123, parse_size('123 B'))
        # Units
        self.assertEqual(1 << 10, parse_size('1KB'))
        self.assertEqual(1 << 20, parse_size('1MB'))
        self.assertEqual(1 << 30, parse_size('1GB'))
        self.assertEqual(1 << 40, parse_size('1TB'))
        self.assertEqual(1 << 50, parse_size('1PB'))
        self.assertEqual(1 << 60, parse_size('1EB'))
        self.assertEqual((1 << 10) * 0.5, parse_size('0.5KB'))
        self.assertEqual((1 << 20) * 0.5, parse_size('0.5MB'))
        self.assertEqual((1 << 30) * 0.5, parse_size('0.5GB'))
        self.assertEqual((1 << 40) * 0.5, parse_size('0.5TB'))
        self.assertEqual((1 << 50) * 0.5, parse_size('0.5PB'))
        self.assertEqual((1 << 60) * 0.5, parse_size('0.5EB'))
        self.assertEqual((1 << 70) * 0.5, parse_size('0.5ZB'))
        self.assertEqual((1 << 80) * 0.5, parse_size('0.5YB'))
        # Unambiguous unit format
        self.assertEqual(123, parse_size('123iB'))
        self.assertEqual(123, parse_size('123iB', assume_binary=False))
        self.assertEqual(1 << 10, parse_size('1KiB'))
        self.assertEqual(1 << 10, parse_size('1kIb'))
        self.assertEqual(1 << 10, parse_size('1KIB'))
        self.assertEqual(1 << 10, parse_size('1KiB', assume_binary=False))
        # Case and spaces
        self.assertEqual(1 << 10, parse_size('1Kb'))
        self.assertEqual(1 << 10, parse_size('1kB'))
        self.assertEqual(1 << 10, parse_size('1kb'))
        self.assertEqual(1 << 10, parse_size('1 kb'))
        self.assertEqual(1 << 10, parse_size('      1kb'))
        self.assertEqual(1 << 10, parse_size('1kb      '))
        self.assertEqual(1 << 10, parse_size('1      kb'))
        # Incorrect input
        self.assertRaises(ValueError, parse_size, (''))
        self.assertRaises(ValueError, parse_size, ('abc'))
        self.assertRaises(ValueError, parse_size, ('12 HB'))
        self.assertRaises(ValueError, parse_size, ('12 BB'))
        self.assertRaises(ValueError, parse_size, ('12 BKB'))
        # Already rendered
        self.assertEqual(123, parse_size(123))
        
        # Non binary
        # Usual case
        self.assertEqual(123, parse_size('123', assume_binary=False))
        self.assertEqual(123, parse_size('123B', assume_binary=False))
        self.assertEqual(123, parse_size('123 B', assume_binary=False))
        # Units
        self.assertEqual(1 * (10 ** 3), parse_size('1KB', assume_binary=False))
        self.assertEqual(1 * (10 ** 6), parse_size('1MB', assume_binary=False))
        self.assertEqual(1 * (10 ** 9), parse_size('1GB', assume_binary=False))
        self.assertEqual(1 * (10 ** 12), parse_size('1TB', assume_binary=False))
        self.assertEqual(1 * (10 ** 15), parse_size('1PB', assume_binary=False))
        self.assertEqual(1 * (10 ** 18), parse_size('1EB', assume_binary=False))
        self.assertEqual((1 * (10 ** 3)) * 0.5, parse_size('0.5KB', assume_binary=False))
        self.assertEqual((1 * (10 ** 6)) * 0.5, parse_size('0.5MB', assume_binary=False))
        self.assertEqual((1 * (10 ** 9)) * 0.5, parse_size('0.5GB', assume_binary=False))
        self.assertEqual((1 * (10 ** 12)) * 0.5, parse_size('0.5TB', assume_binary=False))
        self.assertEqual((1 * (10 ** 15)) * 0.5, parse_size('0.5PB', assume_binary=False))
        self.assertEqual((1 * (10 ** 18)) * 0.5, parse_size('0.5EB', assume_binary=False))
        self.assertEqual((1 * (10 ** 21)) * 0.5, parse_size('0.5ZB', assume_binary=False))
        self.assertEqual((1 * (10 ** 24)) * 0.5, parse_size('0.5YB', assume_binary=False))


class TemplateTagTest(TestCase):

    def test_tag_should_support_none_values(self):
        self.assertEqual('', filesize(None))


quicktest.py
import os
import sys
import argparse
import django
from django.conf import settings

class QuickDjangoTest(object):
    """
    A quick way to run the Django test suite without a fully-configured project.

    Example usage:

        >>> QuickDjangoTest('app1', 'app2')

    Based on a script published by Lukasz Dziedzia at: 
    http://stackoverflow.com/questions/3841725/how-to-launch-tests-for-django-reusable-app
    """
    DIRNAME = os.path.dirname(__file__)

    def __init__(self, *args, **kwargs):
        self.apps = args
        self.run_tests()

    def run_tests(self):
        """
        Fire up the Django test suite developed for version 1.2
        """
        settings.configure(
            DATABASES={
                'default': {
                    'ENGINE':  'django.db.backends.sqlite3',
                    'NAME': os.path.join(self.DIRNAME, 'database.db'),
                    'USER': '',
                    'PASSWORD': '',
                    'HOST': '',
                    'PORT': '',
                }
            },
            INSTALLED_APPS= self.apps,
        )

        # Setup configuration
        django.setup()

        from django.test.runner import DiscoverRunner as Runner

        failures = Runner().run_tests(self.apps, verbosity=1)
        if failures:  # pragma: no cover
            sys.exit(failures)


if __name__ == '__main__':
    """
    What do when the user hits this file from the shell.

    Example usage:

        $ python quicktest.py app1 app2

    """
    parser = argparse.ArgumentParser(
        usage="[args]",
        description="Run Django tests on the provided applications."
    )
    parser.add_argument('apps', nargs='+', type=str)
    args = parser.parse_args()
    QuickDjangoTest(*args.apps)


sizefield/models.py
from django.db import models
from django.core import exceptions
from django.utils.translation import gettext_lazy as _

from sizefield.utils import parse_size
from sizefield.widgets import FileSizeWidget


class FileSizeField(models.BigIntegerField):
    default_error_messages = {
        'invalid': _(u'Incorrect file size format.'),
    }

    def formfield(self, **kwargs):
        kwargs['widget'] = FileSizeWidget
        kwargs['error_messages'] = self.default_error_messages
        return super(FileSizeField, self).formfield(**kwargs)

    def to_python(self, value):
        if value is None:
            return None
        try:
            return parse_size(value)
        except ValueError:
            raise exceptions.ValidationError(self.error_messages['invalid'])


try:
    from south.modelsinspector import add_introspection_rules

    add_introspection_rules([], [r"^sizefield\.models\.FileSizeField"])
except ImportError:
    pass


sizefield/widgets.py
from django import forms
from django.core.validators import EMPTY_VALUES

from sizefield.utils import filesizeformat, parse_size


class FileSizeWidget(forms.TextInput):

    def render(self, name, value, attrs=None, renderer=None):
        if value:
            try:
                value = filesizeformat(value)
            except ValueError:
                pass
        return super(FileSizeWidget, self).render(name, value, attrs, renderer=renderer)

    def value_from_datadict(self, data, files, name):
        """
        Given a dictionary of data and this widget's name, returns the value
        of this widget. Returns None if it's not provided.
        """
        value = super(FileSizeWidget, self).value_from_datadict(data, files, name)
        if value not in EMPTY_VALUES:
            try:
                return parse_size(value)
            except ValueError:
                pass
        return value


sizefield/__init__.py
import re
from decimal import Decimal, getcontext

from django.utils.translation import ugettext as _


FILESIZE_UNITS = {
    _('B') : 2**10, 
    _('KB'): 2**20, 
    _('MB'): 2**30, 
    _('GB'): 2**40, 
    _('TB'): 2**50,
    _('PB'): 2**60,
}


def render_size(value, decimals=1):
    """
    @rtype string
    """
    suffixes = sorted(FILESIZE_UNITS.items(), cmp=lambda x,y: cmp(x[1],y[1]))
    try:
        value = int(value)
    except ValueError:
        # check if value already pretty !
        value = parse_size(value)  # Will raise ValueError if incorrect format

    for suf, lim in suffixes:
        if value >= lim:
            continue
        else:
            value = Decimal(value) / (lim / 2**10)
            value = value.quantize(Decimal(10)**-decimals)  # 10^-n precision.
            return "%(value)s%(suf)s" % locals()


def parse_size(txt):
    """
    @rtype int
    """
    if isinstance(txt, (int,long)):
        return txt
    p = re.compile(r'^(?P<value>[0-9\.]+?)\s*(?P<unit>(B{0,1}|[KMGT]{1}B{1})?)$')
    r = p.match(txt)
    if r:
        value = float(r.group('value'))  # Will raise ValueError if incorrect format
        unit  = r.group('unit') or _('B')
        multi = FILESIZE_UNITS[unit]  # should never fail.
        return int(value * multi/2**10)
    # Regex pattern was not matched
    raise ValueError(_("Size '%s' has incorrect format") % txt)


sizefield/utils.py
from __future__ import unicode_literals

import sys
import re
import operator

from django.utils import formats
from django.utils.translation import gettext as _
from django.conf import settings


if sys.version_info >= (3, 0):
    xrange = range

SIZEFIELD_FORMAT = getattr(settings, 'SIZEFIELD_FORMAT', '{value}{unit}')

file_size_re = re.compile(r'^(?P<value>[0-9\.,]+?)\s*(?P<unit>[KMGTPEZY]?B?)$', re.IGNORECASE)
FILESIZE_UNITS = {
    'B': 1,
    'KB': 1 << 10,
    'MB': 1 << 20,
    'GB': 1 << 30,
    'TB': 1 << 40,
    'PB': 1 << 50,
    'EB': 1 << 60,
    'ZB': 1 << 70,
    'YB': 1 << 80,
}


def filesizeformat(bytes, decimals=1):
    """
    Formats the value like a 'human-readable' file size (i.e. 13 KB, 4.1 MB,
    102 bytes, etc).
    Based on django.template.defaultfilters.filesizeformat
    """

    try:
        bytes = float(bytes)
    except (TypeError, ValueError, UnicodeDecodeError):
        raise ValueError

    def filesize_number_format(value):
        return formats.number_format(round(value, decimals), decimals)

    units_list = sorted(FILESIZE_UNITS.items(), key=operator.itemgetter(1))

    value = unit = None
    len_unints_list = len(units_list)
    for i in xrange(1, len_unints_list):
        if bytes < units_list[i][1]:
            prev_unit = units_list[i - 1]
            value = filesize_number_format(bytes / prev_unit[1])
            unit = prev_unit[0]
            break

    if value is None:
        value = filesize_number_format(bytes / units_list[-1][1])
        unit = units_list[-1][0]

    return SIZEFIELD_FORMAT.format(value=value, unit=unit)


def parse_size(size):
    """
    @rtype int
    """
    if isinstance(size, int):
        return size

    r = file_size_re.match(size.strip())
    if r:
        clean_value = r.group("value").replace(",", ".")
        value = float(clean_value)
        unit = r.group('unit').upper()
        if not unit.endswith('B'):
            unit += 'B'
        return int(value * FILESIZE_UNITS[unit])

    # Regex pattern was not matched
    raise ValueError(_("Size '%s' has incorrect format") % size)
